workspace/src/localization/particle_filter_estimation/particle_filter_estimation/particle_filter_estimation.py

```diff
@@ -0,0 +1,205 @@
+        i_wheel = self.get_parameter("i_wheel").get_parameter_value().double_value
+        self.declare_parameter("gamma", 0.33333333)
+        gamma = self.get_parameter("gamma").get_parameter_value().double_value
+        self.declare_parameter("tau_0", 0.3)
+        tau_0 = self.get_parameter("tau_0").get_parameter_value().double_value
+        self.declare_parameter("omega_0", 30.0)
+        omega_0 = self.get_parameter("omega_0").get_parameter_value().double_value
+        dyn = [c_1, c_0, l, r_wheel, i_wheel, gamma, tau_0, omega_0]
+        self.get_logger().info(str(dyn))
+
+        # update frequency of this node
+        self.freq = 10.0
+
+        self.gps = ""
+        self.ground_truth = ""
+        self.mag = ""
+
+        # x, y, from measurements
+        self.x = 0
+        self.y = 0
+
+        # what we will be using for our state vector. (x, y, theta yaw, v vel)
+        self.state = np.zeros((4, 1))
+        np.vstack(self.state)
+
+        self.init_x = 0.0
+        self.init_y = 0.0
+        self.init_theta = 0.0
+        self.init_v = 0.0
+        self.state[0, 0] = self.init_x
+        self.state[1, 0] = self.init_y
+        self.state[2, 0] = self.init_theta
+        self.state[3, 0] = self.init_v
+
+        self.gps_ready = False
+
+        # ground truth velocity
+        self.gtvy = 0
+        self.gtvx = 0
+        self.D = 0
+
+        # origin, and whether or not the origin has been set yet.
+        self.origin_set = False
+        self.orig_heading_set = False
+
+        # inputs to the vehicle
+        self.throttle = 0.0
+        self.steering = 0
+
+        # time between imu updates, sec
+        self.dt_gps = 1 / self.freq
+
+        # the ROM
+        self.dynamics_model = get_dynamics(self.dt_gps, dyn)
+        # filter
+        self.pf = PF(self.dt_gps, self.dynamics_model)
+
+        # our graph object, for reference frame
+        self.graph = get_coordinate_transfer()
+
+        # subscribers
+        self.sub_gps = self.create_subscription(
+            NavSatFix, "~/input/gps", self.gps_callback, 1
+        )
+
+        self.sub_mag = self.create_subscription(
+            MagneticField, "~/input/magnetometer", self.mag_callback, 1
+        )
+        self.sub_control = self.create_subscription(
+            VehicleInput, "~/input/vehicle_inputs", self.inputs_callback, 1
+        )
+        # publishers
+        self.pub_objects = self.create_publisher(
+            VehicleState, "~/output/filtered_state", 1
+        )
+        self.timer = self.create_timer(1 / self.freq, self.pub_callback)
+
+    # CALLBACKS:
+    def inputs_callback(self, msg):
+        self.inputs = msg
+        self.steering = self.inputs.steering
+        self.throttle = self.inputs.throttle
+
+    def mag_callback(self, msg):
+        self.mag = msg
+        mag_x = self.mag.magnetic_field.x
+        mag_y = self.mag.magnetic_field.y
+        mag_z = self.mag.magnetic_field.z
+        xGauss = mag_x * 0.48828125
```

Comment:
looks pretty similar to ground truth, is it exactly the same?

Reply:
Everything other than the publisher is pretty much the same between all three localization nodes. This is something that could be different for different localization nodes though, so imo makes sense to divide it this way. could see an argument where `ground_truth` gets moved to a different package, something like `pre_process_sensors`, or even just in `sensing`, and then localization algs subscribe to that...